### DISCUSSION TITLE
Two radios on one bus spi/esp32s2

### DISCUSSION THREAD

**Discussion by whoim2** [jgromes/RadioLib#1278] [Q&A]
Im connect first sx1278 module to esp:
```
#define MISO_PIN    37
#define MOSI_PIN    35
#define SCK_PIN     36
#define CS1_PIN     34
#define CS2_PIN     16
#define DIO0_1_PIN  21
#define RST1_PIN    38
#define DIO1_1_PIN  33

SPIClass spi2(FSPI);
SPISettings spiSettings(10000000, MSBFIRST, SPI_MODE0);

SX1278 radio_tx = new Module(CS1_PIN, DIO0_1_PIN, RST1_PIN, DIO1_1_PIN, spi2, spiSettings);

setup:
spi2.begin(SCK_PIN, MISO_PIN, MOSI_PIN, -1);
int state = radio_tx.begin(freq_tx, lora_modes[lora_mode][0], lora_modes[lora_mode][1], lora_modes[lora_mode][2], LORA_SYNC_WORD, pwr, RF_LORA_PRMBL, 0);
check_state(state);
```
return success state code.

Im connect second module to this esp:
```
#define CS2_PIN     16
#define DIO0_2_PIN  14
#define RST2_PIN    13
#define DIO1_2_PIN  4

SX1278 radio_tx = new Module(CS1_PIN, DIO0_1_PIN, RST1_PIN, DIO1_1_PIN, spi2, spiSettings);
SX1278 radio_rx = new Module(CS2_PIN, DIO0_2_PIN, RST2_PIN, DIO1_2_PIN, spi2, spiSettings);

setup:
spi2.begin(SCK_PIN, MISO_PIN, MOSI_PIN, -1);
int state = radio_tx.begin(freq_tx, lora_modes[lora_mode][0], lora_modes[lora_mode][1], lora_modes[lora_mode][2], LORA_SYNC_WORD, pwr, RF_LORA_PRMBL, 0);
  check_state(state);

state = radio_rx.begin(freq_rx, lora_modes[lora_mode][0], lora_modes[lora_mode][1], lora_modes[lora_mode][2], LORA_SYNC_WORD, MIN_PWR, RF_LORA_PRMBL, 0);
  check_state(state);
```

FIRST module have error -16 (spi problem), but second return success!
What am I doing wrong?

**Comment by whoim2** [comment]
im use this and this worked
```
void radio_tx_on(void) {
  digitalWrite(CS2_PIN, HIGH);
  digitalWrite(CS1_PIN, LOW);
}
void radio_rx_on(void) {
  digitalWrite(CS1_PIN, HIGH);
  digitalWrite(CS2_PIN, LOW);
}
void radios_off(void) {
  digitalWrite(CS2_PIN, HIGH);
  digitalWrite(CS1_PIN, HIGH);
}
```

**Comment by jgromes** [comment]
Couple of things:

1. 10 MHz SPI clock you are using is the maximum allowed for SX1278 according to the datasheet. I would try to decrease that a bit.
2. RadioLib handles SPI CS automatically, and only touches the pin directy before and after the SPI transaction. You shouldn't  pull the CS low for prolonged periods of time, some SPI devices really don't like that.
3. What is the output from SPI debug when the -16 error is occuring?

**Comment by whoim2** [comment]
1) try 1,5,8mbit, no effect
2) work with code above, manual cs switches before any radio*. call, and off after.
3) no debug now (

**Reply by jgromes** [reply]
I would strongly suggest trying to get the debug output. Without that, you don't actually know what went wrong , and why manually pulling CS seems to "fix" the issue. -16 is returned when the SPI communication is unreliable, if it was broken completely -2 would have been returned.

**Reply by whoim2** [reply]
> trying to get the debug output

Ok, in monday
